data/transactions/logic/eval.go

```diff
@@ -2802,36 +2777,105 @@ func opAppDeleteGlobalState(cx *evalContext) {
 		return
 	}
 
-	err := cx.appDeleteGlobalKey(key)
+	err := cx.Ledger.DelGlobal(key)
 	if err != nil {
 		cx.err = err
 		return
 	}
 	cx.stack = cx.stack[:last]
 }
 
-func opAssetHoldingGet(cx *evalContext) {
-	last := len(cx.stack) - 1 // asset id
-	prev := last - 1          // account offset
+// We have a difficult naming problem here. In some opcodes, TEAL
+// allows (and used to require) ASAs and Apps to to be referenced by
+// their "index" in an app call txn's foeign-apps or foreign-assets
+// arrays.  That was a small integer, no more than 2 or so, and was
+// often called an "index".  But it was not a basics.AssetIndex or
+// basics.ApplicationIndex.
 
-	assetID := cx.stack[last].Uint
-	accountIdx := cx.stack[prev].Uint
-	fieldIdx := uint64(cx.program[cx.pc+1])
+func appReference(cx *evalContext, ref uint64, foreign bool) (basics.AppIndex, error) {
+	if cx.version >= directRefEnabledVersion {
+		if ref == 0 {
+			return cx.Ledger.ApplicationID(), nil
+		}
+		if ref <= uint64(len(cx.Txn.Txn.ForeignApps)) {
+			return basics.AppIndex(cx.Txn.Txn.ForeignApps[ref-1]), nil
+		}
+		for _, appID := range cx.Txn.Txn.ForeignApps {
+			if appID == basics.AppIndex(ref) {
+				return appID, nil
+			}
+		}
+	} else {
+		// Old rules
+		if foreign {
+			// In old versions, a foreign reference must be an index in ForeignAssets or 0
+			if ref == 0 {
+				return cx.Ledger.ApplicationID(), nil
+			}
+			if ref <= uint64(len(cx.Txn.Txn.ForeignApps)) {
+				return basics.AppIndex(cx.Txn.Txn.ForeignApps[ref-1]), nil
+			}
+		} else {
+			// Otherwise it's direct
+			return basics.AppIndex(ref), nil
+		}
+	}
+	return basics.AppIndex(0), fmt.Errorf("invalid App reference %d", ref)
+}
```

Comment:
This function and the next are the keys to the PR.  They turn a uint64 into the proper basics.A*Index type properly, depending on the versioning of the teal, whether the uint64 is a high or low number, and whether the initial context (in old teal) demanded a direct or indirect reference.